crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -153,30 +154,49 @@ impl BlobVec {
     /// [`BlobVec`]'s `item_layout`
     pub unsafe fn replace_unchecked(&mut self, index: usize, value: OwningPtr<'_>) {
         debug_assert!(index < self.len());
-        // If `drop` panics, then when the collection is dropped during stack unwinding, the
-        // collection's `Drop` impl will call `drop` again for the old value (which is still stored
-        // in the collection), so we get a double drop. To prevent that, we set len to 0 until we're
-        // done.
+
+        let drop = self.drop;
+        let size = self.item_layout.size();
+
+        // Temporarily set the length to zero, so that uninitialized bytes won't
+        // get observed in case this function panics.
         let old_len = self.len;
-        let ptr = self.get_unchecked_mut(index).promote().as_ptr();
         self.len = 0;
-        // Drop the old value, then write back, justifying the promotion
-        // If the drop impl for the old value panics then we run the drop impl for `value` too.
-        if let Some(drop) = self.drop {
-            struct OnDrop<F: FnMut()>(F);
-            impl<F: FnMut()> Drop for OnDrop<F> {
-                fn drop(&mut self) {
-                    (self.0)();
-                }
-            }
-            let value = value.as_ptr();
-            let on_unwind = OnDrop(|| (drop)(OwningPtr::new(NonNull::new_unchecked(value))));
 
-            (drop)(OwningPtr::new(NonNull::new_unchecked(ptr)));
+        // Transfer ownership of the old value out of the vector, so it can be dropped.
+        // SAFETY:
+        // * The caller has promized that `index` fits in this vector.
+        // * The storage location will get overwritten with `value` later, which ensures
+        //   that the element will not get observed or double dropped later.
+        // * If a panic occurs, `self.len` will remain `0`, which ensures a double-drop
+        //   does not occur. Instead, all elements will be forgotten.
+        let old_value = self.get_ptr_mut().byte_add(index * size).promote();
+
+        let dst_addr = old_value.as_ptr();
```

Comment:
Style nit: *_addr is implied by the type already. I think it's ok to omit it and just the classic convention of just `src` and `dst`.